pkg/submitter/submitter.go

```diff
@@ -0,0 +1,504 @@
+	ctx, close := context.WithCancel(ctx)
+	submitter := &Submitter{
+		ctx:       ctx,
+		close:     close,
+		proxy:     proxy,
+		submitter: txSubmitter,
+		cfg:       cfg,
+		resultCh:  make(chan *mining.Result),
+		account:   account,
+		logger:    logger,
+		tellor:    tellor,
+		submitCount: promauto.NewCounter(prometheus.CounterOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_total",
+			Help:        "The total number of submitted solutions",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		}),
+		submitFailCount: promauto.NewCounter(prometheus.CounterOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_fails_total",
+			Help:        "The total number of failed submission",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		}),
+		submitProfit: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_profit",
+			Help:        "The current submit profit in percents",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		},
+			[]string{"slot"},
+		),
+		submitCost: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_cost",
+			Help:        "The current submit cost in 1e18 eth",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		},
+			[]string{"slot"},
+		),
+		submitReward: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_reward",
+			Help:        "The current reward in 1e18 eth",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		},
+			[]string{"slot"},
+		),
+	}
+
+	return submitter, submitter.resultCh
+}
+
+func (s *Submitter) Start() error {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return nil
+		case result := <-s.resultCh:
+			if s.lastSubmitCncl != nil {
+				s.lastSubmitCncl()
+			}
+			var ctx context.Context
+			ctx, s.lastSubmitCncl = context.WithCancel(s.ctx)
+			s.handleSubmit(ctx, result)
+		}
+	}
+}
+
+func (s *Submitter) Stop() {
+	if s.lastSubmitCncl != nil {
+		s.lastSubmitCncl()
+	}
+	s.close()
+	level.Info(s.logger).Log("msg", "submitter shutdown complete")
+}
+
+func (s *Submitter) blockUntilTimeToSubmit(ctx context.Context) {
+	var (
+		lastSubmit time.Duration
+		timestamp  *time.Time
+		err        error
+	)
+	for {
+		select {
+		case <-ctx.Done(): // The context was canceled from the main loop because new work arrived.
+		default:
+		}
+		lastSubmit, timestamp, err = s.lastSubmit()
+		if err != nil {
+			level.Debug(s.logger).Log("msg", "checking last submit time", "err", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		break
+	}
+	var (
+		timeToSubmit context.Context
```

Comment:
```suggestion
		timeToSubmit = context.Background()
```
otherwise it is nill and it panics in 
```
case <-timeToSubmit.Done():
```